qiskit_nature/second_q/mappers/bosonic_mapper.py

```diff
@@ -0,0 +1,58 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2023.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Bosonic Mapper."""
+
+from __future__ import annotations
+
+from qiskit.quantum_info import SparsePauliOp
+from qiskit_nature.exceptions import QiskitNatureError
+
+from qiskit_nature.second_q.operators import BosonicOp
+
+from .qubit_mapper import ListOrDictType, QubitMapper
+
+
+class BosonicMapper(QubitMapper):
+    """Mapper of Bosonic Operator to Qubit Operator
+
+    The following attributes can be read and updated once the ``BosonicMapper`` object
+    has been constructed.
+
+    Attributes:
+        truncation (int): defines the excitation space of the k-th bosonic state. Together with the
+            number of modes required to represent the bosonic operator, it defines the minimum length of
+            the qubit register. The minimum value is 1.
+    """
+
+    def __init__(self, truncation: int) -> None:
+        """
+        Args:
+            truncation: defines the excitation space of the k-th bosonic state. Together with the
+                number of modes required to represent the bosonic operator, it defines the minimum length
+                of the qubit register. The minimum value is 1.
+        """
+        super().__init__()
+        if truncation < 1:
+            raise QiskitNatureError(
+                "Truncation for bosonic linear mapper must be at least 1. "
+                + f"Detected value: {truncation}"
+            )
+        self.truncation = truncation
```

Comment:
A default cannot be set here since "all possible" is an infinite number. Thus, this is a required setting.
And choosing a default on our side would be ill-advised because this can have serious implications on the quality of the result.

Reply:
Ok, with the name stating `truncation (int): defines the excitation space of the k-th bosonic state.` as in it defines something the name truncation still seems a bit weird to me. It feels to me like this is analogous to defining an active space for an electronic structure problem - the goal there being to limit things too, though there at least things are finite to start with.

Reply:
No, this is not related to active spaces in electronic structure. Let me try to explain in as non-technical terms as possible:

Bosonic particles can theoretically be excited an infinite number of times. This is in contrast to fermions where `+_0 +_0` is an invalid operation because a fermion can only be excited once.
Consequently, this means that for bosons `+_0 +_0 +_0 ...` is valid and, thus, one can have states such as `|0>`, `|1>`, `|2>` etc.
In order to map these states onto a qubit register, this linear mapper encodes these in a linear fashion:
```
|0> -> |10000...>
|1> -> |01000...>
|2> -> |00100...>
...
```
As you can see, theoretically we would need an infinite number of qubits to encode all possible bosonic states. That is where this `truncation` comes in, as it specifies the maximum excitation that we allow to be represented. So for example for `truncation=1`, we get:
```
|0> -> |10>
|1> -> |01>
|2> -> can no longer be represented
```
For `truncation=2` we will get out 3 qubits and so on.
This is essentially what the `LinearMapper` does for `SpinOp`, too.

In the future, we should also look into implementing the equivalent of the `LogarithmicMapper`, where one encodes for example:
```
|0> -> |00>
|1> -> |10>
|2> -> |01>
|3> -> |11>
```

Now, the difference to the `LinearMapper` is that in those cases, the `SpinOp` carries the truncation information in the form of its `.spin` attribute. But for the bosonic case, the bosonic operator itself is not the right place to store this, as this is merely an artifact of the mapping procedure.

Reply:
Thanks for the explanation - you still feel truncation is the best name? The field currently says `defines the excitation space` and you explain it as `max excitation` yes it may truncate but the name truncate on it own does not really say what it is just truncating `max_excitation`, `excitation_limit` or something like that seem to be in name more what description says. I guess excitation_truncation, a bit more verbose though. And presently this field is just in the base class right with the parameter docstring being generic about limiting things but that base class is not documented. Can see how it looks after inheritence change I guess.

And maybe this got "lost" in all this discussion, from the first comment above.

>This will only do this check in the constructor here and if you set the attribute directly it won't get done. I would think a better choice, in this case, would be a property setter/getter where the value is checked by the setter. If the setter is also used by the constructor then no matter how the value is set it would always be checked

Reply:
Yes, the property setter is a good idea.

Regarding the name: I really don't see it as a big issue. If we want to make it more verbose, we should probably stick to how this is called in the literature. The main reference here is [Miessen et al.](https://journals.aps.org/prresearch/abstract/10.1103/PhysRevResearch.3.043212) where they talk phrase it as

> The excitation space of the k-th bosonic state will be truncated at a maximum occupation number $n_k^{max}$

That's where the `truncation` name came from originally. But I would be fine with `max_occupation`.
My original explanation kind of misused "excitation" but `+_0` really is a particle creation and thus `|n>` is really not the n-th excited state but rather a bosonic state with n particles in it.

<hr/>

TL;DR: I suggest we rename `truncation -> max_occupation` and that we implement it as a property setter :+1: